controllers/disruption_rollout_controller.go

```diff
@@ -0,0 +1,253 @@
+)
+
+type DisruptionRolloutReconciler struct {
+	Client  client.Client
+	Scheme  *runtime.Scheme
+	BaseLog *zap.SugaredLogger
+	log     *zap.SugaredLogger
+}
+
+func (r *DisruptionRolloutReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	r.log = r.BaseLog.With("disruptionRolloutNamespace", req.Namespace, "disruptionRolloutName", req.Name)
+	r.log.Info("Reconciling DisruptionRollout")
+
+	instance := &chaosv1beta1.DisruptionRollout{}
+	randSource := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	// Fetch DisruptionRollout instance
+	if err := r.Client.Get(ctx, req.NamespacedName, instance); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !instance.DeletionTimestamp.IsZero() {
+		// Add finalizer here if required
+		return ctrl.Result{}, nil
+	}
+
+	// Update the DisruptionRollout status based on the presence of the target resource
+	// If the target resource has been missing for longer than the TargetResourceMissingThreshold, delete the instance
+	targetResourceExists, instanceDeleted, err := r.updateTargetResourcePreviouslyMissing(ctx, instance)
+	if err != nil {
+		// Log error and requeue if status update or deletion fails
+		r.log.Errorw("failed to handle target resource status", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	if instanceDeleted {
+		// Skip reconciliation since the instance has been deleted
+		return ctrl.Result{}, nil
+	}
+
+	disruptions, err := GetChildDisruptions(ctx, r.Client, r.log, instance.Namespace, DisruptionRolloutNameLabel, instance.Name)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Update the DisruptionRollout status with the time when the last disruption was successfully scheduled
+	if err := r.updateLastScheduleTime(ctx, instance, disruptions); err != nil {
+		r.log.Errorw("unable to update LastScheduleTime of DisruptionCron status", "err", err)
+		return ctrl.Result{}, err
+	}
+
+	// Calculate next requeue time
+	requeueAfter := time.Duration(randSource.Intn(5)+15) * time.Second //nolint:gosec
+	requeueTime := requeueAfter.Round(time.Second)
+	scheduledResult := ctrl.Result{RequeueAfter: requeueAfter}
+
+	// Run a new disruption if the following conditions are met:
+	// 1. The target resource is available
+	// 2. The target resource has been updated
+	// 3. The target resource update has not been tested
+	// 4. It's not blocked by another disruption already running
+	// 5. It's not past the deadline
+	if !targetResourceExists {
+		r.log.Infow(fmt.Sprintf("target resource is missing, scheduling next check in %s", requeueTime))
+		return scheduledResult, nil
+	}
+
+	if !r.targetResourceUpdated(&instance.Status) {
+		r.log.Infow("target resource hasn't been modified yet, sleeping")
+		return ctrl.Result{}, nil
+	}
+
+	if instance.Status.LastContainerChangeTime.Before(instance.Status.LastScheduleTime) || instance.Status.LastContainerChangeTime.Equal(instance.Status.LastScheduleTime) {
```

Comment:
Okay, so the watcher is what diffs the hashes, and this is the check we use to ensure we aren't re-testing the same rollout? In that case, we should change the log at L92 to Debug, instead of Info, as we'll be printing it many many times per Rollout

Reply:
Yes, the watcher does hashes comparison and update. The controller is only responsible for scheduling disruptions (if all the conditions are met). This check is ensures that we don't re-test the same thing over and over. 

I have updated the log. Thank you for catching it.